### PR TITLE
docker-daemon: transparently decompress layers in GetBlob

### LINKED ISSUE #192: copy: dest !compression not complied with if source is compressed
This is a bit of a weird issue, but it's something that is exposed if you apply #148. Effectively because `docker-daemon` (and thus `docker/tarfile`) return the set of `DiffIDs` as the `LayerInfos` for a particular schema, you end up with `GetBlob(diffid)` giving you the wrong hash.

I'm not _really_ sure how we should handle this, because I don't think the `docker save`-style archives don't contain enough nicely-formatted information about the compressed layer digests for us to be able to switch. Should we just un-`gzip` layers?

### PR DESCRIPTION
**Pull request by cyphar**
In particular, this ensures that layers are decompressed if the source
layer is compressed but the destination does not permit compression.

This was a particular issue when doing a conversion from oci: to any
transport based on the docker tarfile format (because the Docker tarfile
format references DiffIDs, not the blob digests).

Closes #192
Signed-off-by: Aleksa Sarai <asarai@suse.de>

### CODE REVIEW

**Comment by cyphar** [comment]
We can't really add nice tests for this until #148 is merged, because this is only really exposed when you have a compressed source going to a docker-tarfile source.

**Comment by cyphar** [comment]
However, I am adding tests for this to `skopeo`, so we will be able to test this change.

**Comment by cyphar** [comment]
Eugh, this uncovered another issue. Currently inside `tarfile` we use ` inputInfo.Digest.String()` even when the digest was not given... SIGH.

**Comment by cyphar** [comment]
Okay, so `CopySuite.TestCopyCompression` tests that the old semantics work -- in an attempt to use `dir:` to check that compression was actually done. I'm not really sure how to change this test so that it still tests the same thing it tested previously (that compression is applied).

**Comment by cyphar** [comment]
I've added a `::compress` option to `dir:` transports. It's not a pretty hack, but basically this ensures that we can mirror the old test functionality (though the semantics of what's being tested is slightly different).

**Comment by mtrmac** [comment]
> Eugh, this uncovered another issue. Currently inside `tarfile` we use ` inputInfo.Digest.String()` even when the digest was not given... SIGH.

AFAICT that is just an oversight ( #203 ). Though actually improving the code to handle such a case, as you are doing, would be nicer.

**Comment by mtrmac** [comment]
At the highest level, `ShouldCompressLayers` has been intended to be as is, i.e. one-directional: either compress, or leave in the original format; not as a compress/uncompress toggle.

This is useful e.g. for the remote-repository→`docker-daemon:` copy, where the daemon already can transparently decompress layers, so we should neither compress nor decompress the remote input, only downloading it (and verifying digests/signatures) is perfectly fine. Also…

> I've added a `::compress` option to `dir:` transports. It's not a pretty hack, but basically this ensures that we can mirror the old test functionality (though the semantics of what's being tested is slightly different).

… yeah, basically we’ve been treating `dir:` as a debugging output which records the inputs given as close as possible, and it would be nice to retain this. So that would suggest that an `ImageDestination` might perhaps want a tri-state, compress/decompress/leave unchanged.

But, most importantly, the discussion in #192 on the top-level solution (or perhaps even explaining the problem?) needs to happen first.

File: copy/copy.go [hunk 1]
```diff
@@ -472,19 +472,33 @@ func (ic *imageCopier) copyBlobFromStream(srcStream io.Reader, srcInfo types.Blo
...
-		go compressGoroutine(pipeWriter, destStream) // Closes pipeWriter
-		destStream = pipeReader
+		// Decompress the original stream if it is compressed.
+		if isCompressed {
+			logrus.Debugf("... but first decompressing the original blob")
+			destStream, err = decompressor(destStream)
+			if err != nil {
+				return types.BlobInfo{}, errors.Wrapf(err, "Error decompressing blob %s on-the-fly", srcInfo.Digest)
+			}
+		}
```
Comment (mtrmac) [inline copy/copy.go]: Note that we already might be decompressing the stream once, in the `getOriginalLayerCopyWriter` path above, and doing the same work twice is rather wasteful. OTOH integrating all of that here _is_ close to madness.

(Eventually the `imageCopier` needs to grow good tests. Not today, I guess :) )

File: directory/directory_dest.go [hunk 2]
```diff
@@ -41,7 +41,7 @@ func (d *dirImageDestination) SupportsSignatures() error {
 
 // ShouldCompressLayers returns true iff it is desirable to compress layer blobs written to this destination.
 func (d *dirImageDestination) ShouldCompressLayers() bool {
-	return false
+	return d.ref.compressed
```
Comment (mtrmac) [inline directory/directory_dest.go]: _Really_ not a fan of `dir:` modifying the blobs, because of the the “debugging” / “just tell me what is in the source” purposes.

File: directory/directory_transport.go [hunk 3]
```diff
@@ -58,6 +58,12 @@ type dirReference struct {
 	// (But in general, we make no attempt to be completely safe against concurrent hostile filesystem modifications.)
 	path         string // As specified by the user. May be relative, contain symlinks, etc.
 	resolvedPath string // Absolute path with no symlinks, at least at the time of its creation. Primarily used for policy namespaces.
+
+	// Suffix options.
+	options []string
+
+	// Does the directory want compressed blobs?
+	compressed bool
```
Comment (mtrmac) [inline directory/directory_transport.go]: (review remarks on `dir:` just so that they are not lost if we decided to go this way, not as an endorsement of the concept)

Why have both, and silently accept any invalid input? Merely the `compressed` bool would allow us to both parse and format the cases we care about.

File: directory/directory_transport.go [hunk 4]
```diff
@@ -70,11 +76,44 @@ type dirReference struct {
...
+		if len(parts) != 2 {
+			// Coverage: should never happen
+			return nil, errors.Errorf("dir: invalid path option suffix: %s", path)
+		}
+		options = strings.Split(parts[1], ",")
+		if len(options) == 0 {
+			return nil, errors.Errorf("dir: invalid path option suffix: no options set: %s", path)
+		}
+		path = parts[0]
+	}
```
Comment (mtrmac) [inline directory/directory_transport.go]: If we did do so much parsing for `dir` references, this should probably move to the mode of `ParseReference` doing the string parsing, and `NewReference(path string, compressed bool)` just creating an object (so that API callers can do it cheaply).

File: docker/daemon/daemon_dest.go [hunk 5]
```diff
@@ -157,15 +157,26 @@ func (d *daemonImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobI
...
 		}
 		_, err = streamCopy.Seek(0, os.SEEK_SET)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+
+		if inputInfo.Digest == "" {
+			// In some (hacky) cases a caller might not know the size, but will
+			// know the digest (where "know" == "have a fake digest").
+			inputInfo.Digest = digester.Digest()
```
Comment (mtrmac) [inline docker/daemon/daemon_dest.go]: Actually the whole handling of `Digest` values in this function, especially immediately below, seems to be broken (should be a separate PR) and only happens to work through a random coincidence:

The returned digest must match the file name within the tarball (because `PutManifest` uses the layer digests to compute `layerPaths`. But, right now, we create the file with `inputInfo.Digest` (using, in principle, any algorithm), but we always return a freshly computed `digest.Canonical` value; the two do not generally match.

Basically, if we are provided a digest, we should either use it as is, _and return_, or we need to completely ignore it and compute from scratch (but the rest of the code generally expects us not to do that, for signatures etc.).

As an extension, this also works only because `inputInfo.Digest` is computed using `digest.Canonical`, just like the the computation below; that is, I expect, by design, but the code does not make that obvious (and again, computing twice…)

File: docker/daemon/daemon_dest.go [hunk 6]
```diff
@@ -157,15 +157,26 @@ func (d *daemonImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobI
...
 			return types.BlobInfo{}, err
 		}
 		_, err = streamCopy.Seek(0, os.SEEK_SET)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+
+		if inputInfo.Digest == "" {
+			// In some (hacky) cases a caller might not know the size, but will
+			// know the digest (where "know" == "have a fake digest").
```
Comment (mtrmac) [inline docker/daemon/daemon_dest.go]: When does “having a fake digest” happen? `copy.Image` always clears the digest if it compresses (or, with this PR, decompresses), doesn’t it?

Comment (cyphar) [inline docker/daemon/daemon_dest.go]: Because DiffIDs != Digests, this happens for us (when copying we just have the names of the layers that we're dealing with, but we pretend like we don't have the Digest even though the two should be the same).

I'm going to keep this hunk change, but we can discuss the comment later.

**Comment by cyphar** [comment]
@mtrmac I think this is ready for review now.

**Comment by cyphar** [comment]
@mtrmac @runcom  I think this is ready for review now.

File: copy/fixtures [hunk 7]
```diff
@@ -0,0 +1 @@
+../pkg/compression/fixtures
```
Comment (mtrmac) [inline copy/fixtures]: Using symlinks to deduplicate is fine, but could you please keep the `copy/fixtures` directory and use symlinks to individual files? That would make it easier to add other files to `copy/fixtures` in the future.

File: docker/daemon/daemon_src.go [hunk 8]
```diff
@@ -349,7 +362,28 @@ func (s *daemonImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64,
 		if err != nil {
 			return nil, 0, err
 		}
-		return stream, li.size, nil
+
+		// In order to handle the fact that digests != diffIDs (and thus that a
+		// caller which is trying to verify the blob will run into problems),
+		// we need to decompress blobs. This is a bit ugly, but it's a
+		// consequence of making everything addressable by their DiffID rather
+		// than by their digest...
```
Comment (mtrmac) [inline docker/daemon/daemon_src.go]: “digests != diffIDs” is true in general; it would still be worth explaining explicitly that we (choose to) generate a v2s2 manifest with the available DiffID values, so `GetBlob` needs to return a content matching those DiffID values even if the blob has been compressed.

File: docker/daemon/daemon_src.go [hunk 9]
```diff
@@ -349,7 +362,28 @@ func (s *daemonImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64,
...
+		decompressFunc, reader, err := compression.DetectCompression(stream)
+		if err != nil {
+			return nil, 0, errors.Wrapf(err, "detecting compression in blob %s", info.Digest)
+		}
```
Comment (mtrmac) [inline docker/daemon/daemon_src.go]: Won’t this fail on uncompressed streams, such as those created by `docker save`?

File: docker/daemon/daemon_src.go [hunk 10]
```diff
@@ -349,7 +362,28 @@ func (s *daemonImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64,
...
-		return stream, li.size, nil
+
+		// In order to handle the fact that digests != diffIDs (and thus that a
+		// caller which is trying to verify the blob will run into problems),
+		// we need to decompress blobs. This is a bit ugly, but it's a
+		// consequence of making everything addressable by their DiffID rather
+		// than by their digest...
+		decompressFunc, reader, err := compression.DetectCompression(stream)
+		if err != nil {
+			return nil, 0, errors.Wrapf(err, "detecting compression in blob %s", info.Digest)
```
Comment (mtrmac) [inline docker/daemon/daemon_src.go]: Please capitalize error messages, perhaps wrong but consistent with the rest of the code.

File: docker/daemon/daemon_src.go [hunk 11]
```diff
@@ -349,7 +362,28 @@ func (s *daemonImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64,
...
+		}
+		reader, err = decompressFunc(reader)
+		if err != nil {
+			return nil, 0, errors.Wrapf(err, "decompressing blob %s stream", info.Digest)
+		}
+		newStream := readCloseWrapper{
+			Reader: reader,
+			closeFunc: func() error {
+				return stream.Close()
```
Comment (mtrmac) [inline docker/daemon/daemon_src.go]: Can't this be just `closeFunc: stream.Close`?

File: docker/daemon/daemon_src.go [hunk 12]
```diff
@@ -334,6 +335,18 @@ func (s *daemonImageSource) GetTargetManifest(digest digest.Digest) ([]byte, str
 	return nil, "", errors.Errorf(`Manifest lists are not supported by "docker-daemon:"`)
 }
 
+type readCloseWrapper struct {
+	io.Reader
+	closeFunc func() error
+}
```
Comment (mtrmac) [inline docker/daemon/daemon_src.go]: There already is a `tarReadCloser` in this file. Modifying that to use this `readCloseWrapper` would be nice.

File: docker/daemon/daemon_dest.go (hunk 6 above)
Comment (mtrmac) [inline docker/daemon/daemon_dest.go]: Primarily, is any part of the “handle empty Digest case“ commit still necessary? It seems quite unrelated to the PR topic “transparently decompress layers”.

Secondarily, I still can’t understand what “a fake digest” means in here.

File: docker/daemon/daemon_dest.go (hunk 5 above)
Comment (mtrmac) [inline docker/daemon/daemon_dest.go]: (I think this comment still stands, _if_ anything needs to happen in `PutBlob`.)

File: docker/daemon/daemon_src.go (hunk 9 above)
Comment (cyphar) [inline docker/daemon/daemon_src.go]: No, because "no compression" is considered a valid "compressor" if you look in `pkg/compression`. You only get errors if you get an EOF that was unexpected for example. However, I need to add a check that `decompressorFunc != nil`.

File: docker/daemon/daemon_src.go (hunk 11 above)
Comment (cyphar) [inline docker/daemon/daemon_src.go]: Yes.

**Comment by cyphar** [comment]
> Primarily, is any part of the “handle empty Digest case“ commit still necessary? It seems quite unrelated to the PR topic “transparently decompress layers”.

Sure, but it's in a separate commit and it fixes other issues. Given how long it takes to get anything merged into this repo, I don't feel like creating another pull request for a **single** commit.

File: docker/daemon/daemon_src.go (hunk 12 above)
Comment (cyphar) [inline docker/daemon/daemon_src.go]: The two are wrapping different internal objects. I can't really see a nice way of combining the two.

File: docker/daemon/daemon_dest.go (hunk 6 above)
Comment (cyphar) [inline docker/daemon/daemon_dest.go]: Sure, but it's in a separate commit and it fixes other issues. Given how long it takes to get anything merged into this repo, I don't feel like creating another pull request for a **single** commit.

I can drop the comment if you like. There's no point bike-shedding the comment -- it fixes a concrete issue.

File: docker/daemon/daemon_dest.go (hunk 5 above)
Comment (cyphar) [inline docker/daemon/daemon_dest.go]: Even if it is broken, can we **please** fix it separately. I think the reason this is done this way is because of the whole shitty DiffID code -- which I don't think we can completely fix anyway. I'm just trying to make this code work.

File: docker/daemon/daemon_dest.go (hunk 6 above)
Comment (cyphar) [inline docker/daemon/daemon_dest.go]: Sigh. Okay I'll move this to a separate PR and wait even longer for stuff to get merged.

File: docker/daemon/daemon_dest.go (hunk 5 above)
Comment (cyphar) [inline docker/daemon/daemon_dest.go]: I think it's this way because of the shitty DiffID lie. Not sure if we can properly fix it...

**Comment by cyphar** [comment]
@mtrmac I've moved the `PutBlob` code to a different branch. We can handle that two-line change separately.

File: docker/daemon/daemon_src.go (hunk 12 above)
Comment (mtrmac) [inline docker/daemon/daemon_src.go]: AFAICS the `*tar.Reader` in `tarReadCloser` can just as well be an `io.Reader`; nothing depends on that being a `*.tar.Reader`.

File: docker/daemon/daemon_dest.go (hunk 6 above)
Comment (mtrmac) [inline docker/daemon/daemon_dest.go]: Thanks for splitting this. Smaller PRs have less to block them from getting merged :)

(Really the point is that the API as defined expect the digest to either be unknown or valid, and the rest of this PR should make that true again. If we need change the API for other reasons, sure, let’s talk about that, but _actually explicitly talk about  it_ instead of adding a weird corner case which is unjustified by the rest of the code in the repo.)

**Comment by mtrmac** [comment]
:+1: again for PullApprove

**Comment by cyphar** [comment]
Can we please merge this so that projectatomic/skopeo#280. Not having tests for the OCI and Docker translation code path is **not acceptable** -- at SUSE we actually plan to use this and it's getting frustrating that it keeps BREAKING.

**Comment by cyphar** [comment]
/ping @runcom -- can we please merge this?

**Comment by runcom** [comment]
@cyphar could you rebase and I'll merge?

**Comment by cyphar** [comment]
@runcom Rebase'd.

**Comment by cyphar** [comment]
The build failure makes no sense, I didn't touch `manifest/manifest.go`...

```
$ make deps
go get -t -tags "btrfs_noversion libdm_no_deferred_remove" ./...
# github.com/containers/image/manifest
manifest/manifest.go:57: undefined: v1.MediaTypeImageManifestList
```

**Comment by cyphar** [comment]
Heh, it's broken because of https://github.com/opencontainers/image-spec/commit/a125828918b88b485c45c365ebabcaf34827d366. We aren't vendoring `image-spec` so `go get` fails.

That's just dumb. We should be vendoring things.

**Comment by runcom** [comment]
This is a library, I won't vendor thing and fall into the golang mess of having a library with vendor dirs which clashes with whoever import it. If you're talking about running the test with some vendored code, that may work.

**Comment by cyphar** [comment]
I completely agree that vendoring is a mess, but if you want to run `go` on your library to test it we need to vendor things. Is there a reason that we aren't just doing `test-skopeo` (which should run our unit tests --- right?).

**Comment by runcom** [comment]
> I completely agree that vendoring is a mess, but if you want to run go on your library to test it we need to vendor things. Is there a reason that we aren't just doing test-skopeo (which should run our unit tests --- right?).

not sure `test-skopeo` would run that, if it does, I'm 100% ok doing that instead of go get'ting random libraries from the net.

**Comment by runcom** [comment]
@cyphar to unblock this, can you create a corresponding skopeo PR which vendor this code and run the tests there? After it's green over there I'll merge this. (and sorry for this, we need a better way to handle this, even if it means removing test-skopeo from here if we switch skopeo to not use the latest master every time)

**Comment by cyphar** [comment]
Sure @runcom.

**Comment by cyphar** [comment]
@runcom: projectatomic/skopeo#307

**Comment by cyphar** [comment]
Currently `test-unit-local` doesn't test `vendor/` tests. While this makes sense for most vendors, we should be testing our own code inside skopeo. I can make a PR to improve this situation if you like (I'll do it as part of the PR updating to this patch being merged).

**Comment by cyphar** [comment]
@runcom Tests pass.

**Comment by runcom** [comment]
> Currently test-unit-local doesn't test vendor/ tests. While this makes sense for most vendors, we should be testing our own code inside skopeo. I can make a PR to improve this situation if you like (I'll do it as part of the PR updating to this patch being merged).

:+1: thanks Aleksa

**Comment by runcom** [comment]
LGTM
